iree/compiler/Utils/ModuleUtils.cpp

```diff
@@ -0,0 +1,121 @@
+
+  // Iteratively try suffixes until we find one that isn't used.
+  // It'd be nice if there was a SymbolTable::getUniqueName or something.
+  std::string disambiguatedName;
+  int uniqueingCounter = 0;
+  do {
+    disambiguatedName =
+        llvm::formatv("{0}_{1}", originalName, uniqueingCounter++).str();
+  } while (symbolTable.lookup(disambiguatedName));
+
+  SymbolTableCollection symbolTables;
+  SymbolUserMap symbolUsers(symbolTables, moduleOp);
+  mlir::StringAttr nameAttr =
+      mlir::StringAttr::get(op->getContext(), disambiguatedName);
+  symbolUsers.replaceAllUsesWith(op, nameAttr);
+  SymbolTable::setSymbolName(op, disambiguatedName);
+}
+
+LogicalResult mergeModuleInto(Operation *sourceOp, Operation *targetOp,
+                              SymbolTable &targetSymbolTable,
+                              OpBuilder &targetBuilder) {
+  // Capture source information we need prior to destructively merging.
+  SymbolTable sourceSymbolTable(sourceOp);
+  auto &sourceBlock = sourceOp->getRegion(0).front();
+  auto sourceOps = llvm::to_vector<8>(
+      llvm::map_range(sourceBlock, [&](Operation &op) { return &op; }));
+
+  // Resolve conflicts and move the op.
+  for (auto &sourceOp : sourceOps) {
+    if (sourceOp->hasTrait<OpTrait::IsTerminator>()) continue;
+    if (auto symbolOp = dyn_cast<SymbolOpInterface>(sourceOp)) {
+      auto symbolName = symbolOp.getName();
+
+      // Resolve symbol name conflicts.
+      if (auto targetOp = targetSymbolTable.lookup(symbolName)) {
+        if (symbolOp.getVisibility() == SymbolTable::Visibility::Private) {
+          // Private symbols can be safely folded into duplicates or renamed.
+          if (OperationEquivalence::isEquivalentTo(
+                  targetOp, sourceOp, OperationEquivalence::exactValueMatch,
+                  OperationEquivalence::exactValueMatch,
+                  OperationEquivalence::Flags::IgnoreLocations)) {
+            // Optimization: skip over duplicate private symbols.
+            // We could let CSE do this later, but we may as well check here.
+            continue;
+          } else {
+            // Preserve the op but give it a unique name.
+            renameWithDisambiguatedName(sourceOp, sourceOp, sourceSymbolTable);
+          }
+        } else {
+          // The source symbol has 'nested' or 'public' visibility.
+          if (SymbolTable::getSymbolVisibility(targetOp) !=
+              SymbolTable::Visibility::Private) {
+            // Oops! Both symbols are public and we can't safely rename either.
+            // If you hit this with ops that you think are safe to rename, mark
+            // them private.
+            //
+            // Note: we could also skip linking between executables with
+            // conflicting symbol names. We think such conflicts will be better
+            // fixed in other ways, so we'll emit an error until we find a case
+            // where that isn't true.
```

Comment:
this comment was specific to the executable linking code, could trim or remove it